src/index.tsx

```diff
@@ -138,23 +140,36 @@ const VisGraph = forwardRef<
     if (isEqual(graph.nodes, prevNodes.current)) {
       return; // No change!
     }
-    const { added, removed, updated } = diff(prevNodes.current, graph.nodes);
+    const { added, removed, updated } = diff(nodes.get(), graph.nodes);
 
-    nodes.remove(removed);
-    nodes.add(added);
-    nodes.update(updated);
+    if (removed.length) {
+      nodes.remove(removed);
+    }
+    // Shallow clone nodes to ensure props aren't mutated!
+    if (added.length) {
+      nodes.add(shallowClone(added));
+    }
+    if (updated.length) {
+      nodes.update(shallowClone(updated));
+    }
     prevNodes.current = graph.nodes;
   }, [graph.nodes, nodes]);
 
   useEffect(() => {
     if (isEqual(graph.edges, prevEdges.current)) {
       return; // No change!
     }
-    const { added, removed, updated } = diff(prevEdges.current, graph.edges);
-
-    edges.remove(removed);
-    edges.add(added);
-    edges.update(updated);
+    const { added, removed, updated } = diff(edges.get(), graph.edges);
+    if (removed.length) {
+      edges.remove(removed);
```

Comment:
I would just
```
edges.remove([...removed])
```

Reply:
And correct me if I'm wrong but the if statement seems also unnecessarry, will something happen if we call `add, update, remove` with empty array?

Reply:
Nothing happens if we call them with an empty array, so yeah, I can remove those if statements.
No need for `edges.remove([...removed])` as the remove function doesn't appear to mutate the inputs, plus if it's going in removed, it by definition came from the mutable `edges` source.

Reply:
But do we need that `shallowClone` array or is `[...array]` enough for it?

Reply:
For everything but the remove function, yes. we need the shallowClone, because it mutates the array element objects by adding an `id` onto it if there isn't one, which goes very much against how react works, so in order to protect the users of this component, it's best to shallow clone the individual elements of the array.